src/semidiscretization/semidiscretization_coupled.jl

```diff
@@ -0,0 +1,366 @@
+end
+
+
+@inline Base.ndims(semi::SemidiscretizationCoupled) = ndims(semi.semis[1])
+
+@inline nmeshes(semi::SemidiscretizationCoupled) = length(semi.semis)
+
+@inline Base.real(semi::SemidiscretizationCoupled) = promote_type(real.(semi.semis)...)
+
+@inline Base.eltype(semi::SemidiscretizationCoupled) = promote_type(eltype.(semi.semis)...)
+
+@inline function ndofs(semi::SemidiscretizationCoupled)
+  sum(ndofs, semi.semis)
+end
+
+@inline function polydeg(semi::SemidiscretizationCoupled)
+  _, _, solver, _ = mesh_equations_solver_cache(semi.semis[1])
+
+  polydeg(solver)
+end
+
+@inline function nelements(semi::SemidiscretizationCoupled)
+  return sum(semi.semis) do semi_
+    mesh, equations, solver, cache = mesh_equations_solver_cache(semi_)
+
+    nelements(mesh, solver, cache)
+  end
+end
+
+
+@inline function mesh_equations_solver_cache(semi::SemidiscretizationCoupled)
+  _, equations, _, _ = mesh_equations_solver_cache(semi.semis[1])
+  return nothing, equations, nothing, nothing
+end
+
+
+function calc_error_norms(func, u_ode::AbstractVector, t, analyzers,
+                          semi::SemidiscretizationCoupled, caches_analysis;
+                          normalize=true)
+  @unpack semis, u_indices = semi
+
+  # Sum up L2 integrals, use max on Linf error
+  op(x, y) = (x[1] + y[1], max(x[2], y[2]))
+
+  l2_error, linf_error = mapreduce(op, 1:nmeshes(semi)) do i
+    calc_error_norms(func, u_ode[u_indices[i]], t, analyzers[i],
+                     semis[i], caches_analysis[i]; normalize=false)
+  end
+
+  if normalize
+    # For L2 error, divide by total volume
+    total_volume_ = total_volume(semi)
+    l2_error = @. sqrt(l2_error / total_volume_)
+  end
+
+  return l2_error, linf_error
+end
+
+
+function integrate(func::Func, u_ode::AbstractVector, semi::SemidiscretizationCoupled; normalize=true) where {Func}
+  @unpack semis, u_indices = semi
+
+  integral = sum(1:nmeshes(semi)) do i
+    mesh, equations, solver, cache = mesh_equations_solver_cache(semis[i])
+    u = wrap_array(u_ode[u_indices[i]], mesh, equations, solver, cache)
+
+    integrate(func, u, mesh, equations, solver, cache, normalize=false)
+  end
+
+  # Normalize with total volume
+  if normalize
+    total_volume_ = total_volume(semi)
+    integral = integral / total_volume_
+  end
+
+  return integral
+end
+
+
+function total_volume(semi::SemidiscretizationCoupled)
+  sum(semi.semis) do semi_
+    mesh, equations, solver, cache = mesh_equations_solver_cache(semi_)
+    total_volume(mesh, solver, cache)
+  end
+end
+
+
+function compute_coefficients(t, semi::SemidiscretizationCoupled)
+  @unpack u_indices = semi
+
+  u_ode = Vector{real(semi)}(undef, u_indices[end][end])
+
+  for i in 1:nmeshes(semi)
+    # Call `compute_coefficients` in `src/semidiscretization/semidiscretization.jl`
+    u_ode[u_indices[i]] .= compute_coefficients(t, semi.semis[i])
+  end
+
+  return u_ode
+end
+
+
+function allocate_coupled_boundary_conditions(boundary_conditions, semi) end
```

Comment:
Why define this?

Reply:
This will be called for every boundary condition, but we only want to do something for `BoundaryConditionCoupled`.

Reply:
Then you should put this as a comment next to it (and also to all the other places where I left a similar comment and where this applies).